### ISSUE TITLE
Unable to catch/suppress getDownloadURL 404 error

### ISSUE THREAD

**Issue by jjstanton** [firebase/firebase-js-sdk#6969]
### Describe your environment

  * Operating System version: Win 10 Pro 22H2
  * Browser version: Chrome 109.0.5414.75
  * Firebase SDK version: 9.15.0
  * Firebase Product: storage

#### Steps to reproduce:

When using the `getDownloadURL` method and the specified file is not in storage, a 404 error is thrown. I am not able to catch the error or prevent it from being thrown, which leads to a very messy console when iterating through hundreds of images.

Someone else has mentioned this issue on [stack overflow](https://stackoverflow.com/questions/69396848/how-to-catch-a-firebase-storage-error-when-the-file-doesnt-exist) as well.

#### Relevant Code:

```javascript
   const fbStorageRef = ref(storage, `images/${ID}/${ID}.png`);

    try {
      getDownloadURL(fbStorageRef) //<--- 404 thrown within method
      .then( response => {
          // Never gets here
      })
      .catch( error => {
          // Never gets here
      });
    } catch (e) {
      // Never gets here
    }
```

**Comment by maneesht** [comment]
@jjstanton My assumption is that the stack overflow is referring to the default error that Chrome always logs whenever an HTTP request is unsuccessful, which will happen regardless of whether you catch the error or not. 
As far as your code, I was not able to reproduce your issue. When the object is not found, we throw a `FirebaseError: Firebase Storage: Object <ref-location> does not exist. (storage/object-not-found)` error, which I was able to catch with your code. As far as the GET 404 that prints out, that's due to Chrome recognizing that a request was unsuccessful, and not from Firebase's thrown error.

**Comment by maneesht** [comment]
Please let me know if I can help any further.

**Comment by maneesht** [comment]
Can you print out the full error that gets printed?

**Comment by jjstanton** [comment]
@maneesht Thanks for the assistance. I see what you mean by chrome throwing the 404 error. 

Here is the error I am getting for each missing file: 

`GET https://firebasestorage.googleapis.com/v0/b/[my-app-domain]/o/club-images%2F119%2F119-24x24.png 404`

Perhaps I am assuming wrong but I thought the `getDownloadURL` method would only provide the URL for the file in storage, not fetch the file or which ever event is causing the 404. How can I prevent the error prior to using this method?

**Comment by maneesht** [comment]
@jjstanton - so in order to fetch the download url, we need to get the metadata for the specific file. That's what that GET does. As far as what you can do to prevent that error from showing up, there really isn't any way to suppress it besides filtering out errors in your console, as we do not have control over what Chrome does whenever an HTTP request fails.

But, the catch in your promise is correctly catching the error FirebaseStorage throws, otherwise you'd have an additional error show in the console.

**Comment by jjstanton** [comment]
@maneesht I understand what you're saying but it feels like there is a need for a `exists` method here to prevent a fetch if I don't sure if the file exists, thus preventing the fetch and the error to begin with. I know the file does not exist, yet I still don't get a `storage/object-not-found` error of any sort, only a failed get request via chrome in the console and in the network tab:

`error: {code: 404, message: "Not Found."}`


I've managed to prevent the error using the `listAll` method but I'm not sure if I will get a negative performance hit for using it.

```javascript
    const fbStorageListRef = storageRef(storage, `some-folder/${someID}`);
    const fbStorageFileRef = storageRef(storage, `some-folder/${someID}/some-file-name.png`);

    try {
      listAll(fbStorageListRef)
        .then( (list) => {
          if (list.items.length > 0) {
            getDownloadURL(fbStorageFileRef)
              .then( (url) => {
                //Use url here
              }).catch( error => {
                //url error
              });
          }
      })
    } catch (error) {
      //try-catch error
    }
```

**Comment by maneesht** [comment]
@jjstanton - the only way for us to know whether the file exists or not is if we make an HTTP request out to the server. Even if we had some kind of an `exists()` function, it would do the same thing. 

In your previous example, it doesn't look like you're logging the error, is that part just missing from your code? 

And `listAll` is actually a great way of getting the list of files if you know the path in which the files are contained.

**Comment by jjstanton** [comment]
@maneesht - I'm aware there would be a HTTP request but imagined there would be an easier method like `exists()` or perhaps if the `getDownloadUrl` could do what I'm doing with the `listAll`. I've been fetching against a bucket of about 70k folders, so this has been driving me mad.

In the examples above, I excluded the logging of the error.

Seems like I've solved the issue with `listAll`, so I'm happy to close the issue. Thanks for the assistance.
